### PR TITLE
Fixes #5500 - Add a GitHub Actions workflow that builds the docs

### LINKED ISSUE #5500: Automate build and publish of the user guide
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The main datafusion documentation site at https://arrow.apache.org/datafusion is good because:
1. They look good and follow the best practice for open source project documentation 
2. They are associated clearly with the overall arrow.apache.org 

The documentation source is at https://github.com/apache/arrow-datafusion/tree/main/docs. 

https://arrow.apache.org/datafusion is updated (typically by @andygrove) when new releases of arrow-datafusion are published to crates.io (for example,  https://github.com/apache/arrow-site/pull/313)

However, the current setup has a few notable issues:
1. It is behind what is in the repo as it is only updated every release (every 2 weeks at the time of writing)
2. The content of the landing page in the repo (README): https://github.com/apache/arrow-datafusion has diverged from the user guide (as I believe developers want the latest content and the only way to see that without building the docs locally is README.md)
4. The manual update process is somewhat cumbersome

I think if the user guide was more immediately updated, people would be more likely to contribute to it as well.

**Describe the solution you'd like**
I would like some mechanism to see the latest, rendered version of the user guide as a webpage. 

- [ ] On every commit to main, the site would be updated with the latest version of the user guide
- [ ] The main README.md page in arrow-datafusion would redirect to the hosted site

Bonus points for 

**Describe alternatives you've considered**
Perhaps we could make a github pages site https://pages.github.com/ ? 

**Additional context**

### PR DESCRIPTION
**Pull request by martin-g**
# Which issue does this PR close?

Closes #5500 

# Rationale for this change

Deploy the Datafusion website whenever a change is made in the `docs/` folder.

# What changes are included in this PR?

A new Github Action is introduced that builds the docs and copies them to `asf-site` branch

# Are these changes tested?

Yes, I have tested them in my fork.

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by martin-g** [comment]
TODOs for the Datafusion committers:
1) create `asf-site` branch in `arrow-datafusion` repo
2) delete the `datafusion` folder in `arrow-site` repo

**Comment by martin-g** [comment]
CC @kou @alamb

File: .github/workflows/docs.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+on:
+  push:
+    branches:
+      - master
```
Comment (kou) [inline .github/workflows/docs.yaml:4]:
```suggestion
      - main
```

BTW, is it intentional to publish the `main` branch contents (not tagged contents) to https://arrow.apache.org/datafusion/ ?

FYI: ADBC publishes multiple versions:

* https://arrow.apache.org/adbc/ : Redirects to the latest release
* https://arrow.apache.org/adbc/0.2.0/ : The latest release
* https://arrow.apache.org/adbc/0.1.0/ : An old release
* https://arrow.apache.org/adbc/main/ : The next release (not released yet)

See also:
* https://github.com/apache/arrow-adbc/blob/main/.github/workflows/nightly-website.yml
* https://github.com/apache/arrow-adbc/blob/main/ci/scripts/website_build.sh

File: .github/workflows/docs.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+      - name: Copy & push the generated HTML
+        run: |
+          set -x
+          cd asf-site/
+          # delete anything but: 1) '.'; 2) '..'; 3) .git/
+          rm -rf .github .buildinfo .gitattributes .github_changelog_generator .gitignore .gitmodules
+          find ./ | grep -vE "^./$|^../$|^./.git" | xargs rm -rf
+          cp ../.asf.yaml .
+          cp -rT ../docs/build/html/ .
```
Comment (kou) [inline .github/workflows/docs.yaml:52]: FYI: We can use `rsync` for this like https://github.com/apache/arrow-site/blob/main/.github/workflows/deploy.yml#L98-L107 :

```suggestion
          rsync \
            -a \
            --delete \
            --exclude '/.git/' \
            ../docs/build/html/ \
            ./
          cp ../.asf.yaml .
```

File: .asf.yaml [hunk 3]
```diff
@@ -39,3 +39,7 @@ github:
     rebase: false    
   features:
     issues: true        
+
+publish:
+  whoami: asf-site
+  subdir: datafusion
```
Comment (kou) [inline .asf.yaml:45]: Could you add a new line here?

**Comment by kou** [comment]
> 1. create `asf-site` branch in `arrow-datafusion` repo

Done.

File: .asf.yaml (hunk 3 above)
Comment (martin-g) [reply .asf.yaml]: Done!

**Comment by alamb** [comment]
I tested this PR out in my fork and it works great! Thank you @martin-g 

Here https://github.com/alamb/arrow-datafusion/tree/asf-site

What are the next steps after this PR to actually get this site served?

On my fork I configured github pages:

Screenshot 2023-03-21 at 4 42 13 PM

And it published https://github.com/alamb/arrow-datafusion/actions/runs/4483635485 which is cool

But the site looks pretty bad https://alamb.github.io/arrow-datafusion

Screenshot 2023-03-21 at 5 00 56 PM

**Comment by alamb** [comment]
> 2. delete the datafusion folder in arrow-site repo

Here is a PR to do this https://github.com/apache/arrow-site/pull/336

**Comment by alamb** [comment]
However, when I check out the site locally and open it, it works great 👍 

So I think that means the issue is with how github pages is publishing it

Screenshot 2023-03-21 at 5 10 53 PM

File: .github/workflows/docs.yaml (hunk 1 above)
Comment (kou) [reply .github/workflows/docs.yaml]: @alamb Do you have a preference of this?

**Comment by kou** [comment]
We need to add `.nojekyll` to the `asf-site` like https://github.com/apache/arrow-site/blob/main/.github/workflows/deploy.yml#L112 to use GitHub Pages for preview.
If we don't have `.nojekyll`, our `asf-site` contents are processed by Jekyll. Jekyll ignores `_*`. So `_static/styles/*` aren't published on GitHub Pages.

**Review by kou — APPROVED** [review]
+1

**Comment by kou** [comment]
> It would be nice to sort out how this content will be hosted (though maybe I just don't understand how the process works)

ASF provides https://arrow.apache.org/ (not GitHub Pages) that serves the contents in https://github.com/apache/arrow-site/tree/asf-site .
With the `.asf.yaml` configuration in this PR, https://arrow.apache.org/datafusion/ uses https://github.com/apache/arrow-datafusion/tree/asf-site instead of https://github.com/apache/arrow-site/tree/asf-site/datafusion .

**Comment by alamb** [comment]
I tested the latest and greatest code on my fork and it seems to have worked great! https://alamb.github.io/arrow-datafusion/

Thank you so much @martin-g  and @kou 

Screenshot 2023-03-22 at 6 28 42 AM

I am going to merge this PR and see how it goes!

**Comment by alamb** [comment]
My test shows that the process is working : https://github.com/apache/arrow-datafusion/issues/5500#issuecomment-1479366153 

Thanks again @kou  and @martin-g  -- this is really great.

**Comment by alamb** [comment]
PR to update release process: https://github.com/apache/arrow-datafusion/pull/5687

File: docs/source/_templates/docs-sidebar.html [hunk 4]
```diff
@@ -1,7 +1,4 @@
 
-<a class="navbar-brand" href="{{ pathto(master_doc) }}">
```
Comment (alamb) [inline docs/source/_templates/docs-sidebar.html:2]: Here is a proposal to get the logo back : https://github.com/apache/arrow-datafusion/pull/5688

**Pull request merged**